### PR TITLE
Nftdictionary

### PR DESCRIPTION
**Pull request by admazzola**
Adds the NFTDictionary contract to the project.  This serves as a data-store for metadata related to TellerNFTv1.  

Upon deployment, mass data is uploaded into the NFTDictionary contract.  When a new TellerNFT is claimed (minted) then its data is added to the NFTDictionary.

### CODE REVIEW

File: contracts/nft/TellerNFTDictionary.sol [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        baseLoanSizes[index] = newTier.baseLoanSize;
+        hashes[index] = newTier.hashes;
+        contributionAssets[index] = newTier.contributionAsset;
+        contributionSizes[index] = newTier.contributionSize;
+        contributionMultipliers[index] = newTier.contributionMultiplier;
+
+        return true;
+    }
+
+    function setAllTokenTierMappings(uint256[] memory tiersMapping)
```
Comment (passabilities) [inline contracts/nft/TellerNFTDictionary.sol:113]: @admazzola please check the rest of the functions and add comments 🙏🏼

File: contracts/nft/TellerNFTDictionary.sol [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+    /**
+        New methods for the dictionary
+    */
+
+    /**
+     * @notice It returns information about a Tier for a token ID.
+     * @param tokenId ID of the token to get Tier info.
+     */
+    function tokenBaseLoanSize(uint256 tokenId)
```
Comment (passabilities) [inline contracts/nft/TellerNFTDictionary.sol:197]: Incorrect doc strings for this and functions below

File: contracts/nft/distributor/entry/claim.sol [hunk 3]
```diff
@@ -34,6 +37,8 @@ contract ent_claim_NFTDistributor_v1 is
                 "TellerNFT Distributor: invalid proof"
             );
 
+            ///uint256[] preOwnedTokens = distributorStore().nft.getOwnedTokens(account);
```
Comment (passabilities) [inline contracts/nft/distributor/entry/claim.sol:40]: can remove

File: contracts/nft/distributor/entry/claim.sol [hunk 4]
```diff
@@ -42,8 +47,32 @@ contract ent_claim_NFTDistributor_v1 is
...
+            uint256 offset = postOwnedTokens.length - requests[i].amount;
+
+            for (uint256 k; k < requests[i].amount; k++) {
+                newlyMintedTokenIds[k] = postOwnedTokens[k + offset];
+            }
+
+            distributorStore().dictionary.setTokenTierForTokenIds(
+                newlyMintedTokenIds,
+                tierIndex
+            );
```
Comment (passabilities) [inline contracts/nft/distributor/entry/claim.sol:66]: adding inline comments here would be super beneficial when we come back to this at a later point

File: tasks/fork.ts [hunk 5]
```diff
@@ -21,6 +21,8 @@ export const forkNetwork: ActionType<NetworkArgs> = async (
   const { chain, block, onlyDeployment } = args
   const { log, run, network } = hre
 
+  console.log('forking things', network)
```
Comment (passabilities) [inline tasks/fork.ts:24]: can remove

File: test/integration/dictionary.test.ts [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    let deployer: Signer
+    let diamond: ITellerDiamond
+
+    before(async () => {
+      // Get funded market with NFT
+      /* ;({ diamond } = await fundedMarket({
+        assetSym: market.lendingToken,
+        amount: 100000,
+        tags: ['nft'],
+      }))*/
```
Comment (passabilities) [inline test/integration/dictionary.test.ts:38]: can remove

File: test/integration/dictionary.test.ts [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+              {
+                merkleIndex: merkleIndex + existingMerkles.length,
+                nodeIndex: info.claims[borrower].index,
+                amount: info.claims[borrower].amount,
+                merkleProof: info.claims[borrower].proof,
+              },
+            ])
+            .then(({ wait }) => wait())
+
+          //await claimNFT({ account: borrower, merkleIndex:  merkleIndex+existingMerkles.length }, hre)
```
Comment (passabilities) [inline test/integration/dictionary.test.ts:165]: can remove

File: test/integration/dictionary.test.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            'TellerNFTDictionary'
+          )
+
+          const tokenTier = await nftDictionary.getTokenTierIndex(
+            lastOwnedTokenId
+          )
+
+          console.log(`token tier for ${lastOwnedTokenId} is ${tokenTier}`)
+
+          console.log(`merkleData.tierIndex was ${merkleData.tierIndex}`)
```
Comment (passabilities) [inline test/integration/dictionary.test.ts:182]: can remove

File: test/integration/dictionary.test.ts [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            .connect(deployer)
+            .addMerkle(info.tierIndex, info.merkleRoot)
+            .then(({ wait }) => wait())
+
+          /*const borrower = ethers.utils.getAddress(
+            merkleData.balances[0].address
+          )*/
+
+          console.log(`info claims ${JSON.stringify(info.claims[borrower])}`)
+          console.log(`merkle index ${merkleIndex + existingMerkles.length}`)
```
Comment (passabilities) [inline test/integration/dictionary.test.ts:150]: can remove

File: tasks/nft/add-nft-tiers.ts [hunk 10]
```diff
@@ -91,6 +126,63 @@ const logTier = (tier: TierInfo, indent: number): void => {
...
+
+
+  */
```
Comment (passabilities) [inline tasks/nft/add-nft-tiers.ts:134]: can remove

File: test/unit/nft-dictionary.test.ts [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+    // 0x0021020100030201000302010003020100030201000302010003020100030201
+
+    //BINARY
+    // 100001000000100000000100000000000000110000001000000001000000000000001100000010000000010000000000000011000000100000000100000000000000110000001000000001000000000000001100000010000000010000000000000011000000100000000100000000000000110000001000000001
+
+    expect(results).to.equal(3);
+  
+
+
+  });*/
```
Comment (passabilities) [inline test/unit/nft-dictionary.test.ts:38]: needs to be implemented?

File: test/unit/nft-dictionary.test.ts [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+      0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0,
+      3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3,
+      2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2,
+      1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1,
+      0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0,
+      3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3, 2, 1, 0, 3,
+      2, 1,
+    ]
+
+    console.log('tokenTiers length', tokenTiers.length)
```
Comment (passabilities) [inline test/unit/nft-dictionary.test.ts:106]: can remove comments in file

File: utils/deploy-helpers.ts [hunk 13]
```diff
@@ -42,10 +42,11 @@ export const deploy = async <C extends Contract>(
     const contractName = `${args.contract}${args.mock ? 'Mock' : ''}`
 
     const result = await deploy(contractDeployName, {
+      ...args,
       contract: contractName,
-      libraries: args.libraries,
+      //libraries: args.libraries,
       from: deployer,
-      args: args.args,
+      //args: args.args,
```
Comment (passabilities) [inline utils/deploy-helpers.ts:49]: can remove

**Review by passabilities — CHANGES_REQUESTED** [review]

**Comment by admazzola** [comment]
Making a new pull request..

**Pull request closed without merging**